src/ore/src/cgroup.rs

```diff
@@ -0,0 +1,251 @@
+        let subsystems = fields
+            .nth(1)?
+            .split(',')
+            .filter(|s| !s.is_empty())
+            .map(|s| s.to_owned())
+            .collect();
+        let root = PathBuf::from(fields.next()?);
+        Some(CgroupEntry { subsystems, root })
+    }
+}
+
+/// Parses /proc/self/cgroup into a Vec<CgroupEntry>, if the file exists.
+pub fn parse_proc_self_cgroup() -> Option<Vec<CgroupEntry>> {
+    let file = File::open("/proc/self/cgroup").ok()?;
+    let file = BufReader::new(file);
+    Some(
+        file.lines()
+            .filter_map(|result| result.ok())
+            .filter_map(CgroupEntry::from_line)
+            .collect(),
+    )
+}
+
+/// An entry in /proc/self/mountinfo
+#[derive(Debug, PartialEq)]
+pub struct Mountinfo {
+    root: PathBuf,
+    mount_point: PathBuf,
+    fs_type: String,
+    super_opts: Vec<String>,
+}
+
+impl Mountinfo {
+    fn from_line(line: String) -> Option<Mountinfo> {
+        // https://www.kernel.org/doc/Documentation/filesystems/proc.txt
+        let mut split = line.split(" - ");
+
+        let mut mount_fields = split.next()?.split(' ');
+        let root = PathBuf::from(mount_fields.nth(3)?);
+        let mount_point = PathBuf::from(mount_fields.next()?);
+
+        let mut fs_fields = split.next()?.split(' ');
+
+        let fs_type = fs_fields.next()?.split('.').next()?.to_owned();
+        let super_opts: Vec<String> = fs_fields
+            .nth(1)?
+            .split(',')
+            .filter(|s| !s.is_empty())
+            .map(|s| s.to_owned())
+            .collect();
+
+        Some(Mountinfo {
+            root,
+            mount_point,
+            fs_type,
+            super_opts,
+        })
+    }
+}
+
+/// Parses /proc/self/mountinfo into vectors of Mountinfo objects,
+/// returning (v2_mounts, v1_mounts).
+pub fn parse_proc_self_mountinfo() -> Option<(Vec<Mountinfo>, Vec<Mountinfo>)> {
+    let file = File::open("/proc/self/mountinfo").ok()?;
+    let file = BufReader::new(file);
+    Some(
+        file.lines()
+            .filter_map(|result| result.ok())
+            .filter_map(Mountinfo::from_line)
+            .filter(|mi| mi.fs_type == "cgroup" || mi.fs_type == "cgroup2")
+            .partition(|mi| mi.fs_type == "cgroup2"),
+    )
+}
+
+/// Represents a cgroup memory limit, with both ram and swap maximums if they exist.
+#[derive(Debug)]
+pub struct MemoryLimit {
+    /// Maximum RAM limit, in bytes.
+    pub max: Option<usize>,
+    /// Maximum swap limit, in bytes.
+    pub swap_max: Option<usize>,
+}
+
+fn read_file_to_usize<P: AsRef<Path>>(path: P) -> Option<usize> {
+    match std::fs::read_to_string(path.as_ref()) {
+        Ok(s) => s.trim().parse::<usize>().ok(),
+        Err(_) => None,
+    }
+}
+
+/// Finds the mountpoint corresponding to the provided cgroup v2,
+/// and reads the memory limits within.
+pub fn read_v2_memory_limit(cgroups: &[CgroupEntry], mounts: &[Mountinfo]) -> Option<MemoryLimit> {
```

Comment:
It doesn't look like this and the v1 method below need to be `pub`?

Reply:
Done